script: Start replacing `time` with `std::time` and `chrono`

augustebaum opened the pull request:
This change is intended to be reviewed commit-by-commit.

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes contribute to #30150
- [x] These changes do not require tests because they do not change behavior.

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

augustebaum committed:
Replace `time` with `chrono` in `script/dom/workerglobalscope.rs`

Signed-off-by: Auguste Baum <auguste.apple@gmail.com>

augustebaum committed:
Replace `time` with `std::time` in `script/dom/workerglobalscope.rs`

Signed-off-by: Auguste Baum <auguste.apple@gmail.com>

augustebaum committed:
Replace `time` with `std::time` and `chrono` in `script/script_thread.rs`

Signed-off-by: Auguste Baum <auguste.apple@gmail.com>

augustebaum committed:
Replace `time` with `std::time` in `script/script_thread.rs`

Signed-off-by: Auguste Baum <auguste.apple@gmail.com>

augustebaum committed:
Replace `time` with `std::time` in `script/script_runtime.rs`

Signed-off-by: Auguste Baum <auguste.apple@gmail.com>

augustebaum committed:
Replace `time` with `std::time` in `script/script_runtime.rs`

Signed-off-by: Auguste Baum <auguste.apple@gmail.com>

augustebaum committed:
Replace `time` with `std::time` in `script/dom/htmlmedialelement.rs`

Signed-off-by: Auguste Baum <auguste.apple@gmail.com>

augustebaum committed:
Replace `time` with `std::time` in `script/dom/htmlmedialelement.rs`

Signed-off-by: Auguste Baum <auguste.apple@gmail.com>

augustebaum committed:
Replace `time` with `chrono` in `script/dom/globalscope.rs`

Signed-off-by: Auguste Baum <auguste.apple@gmail.com>

augustebaum committed:
Replace `time` with `std::time` in `script/dom/globalscope.rs`

Signed-off-by: Auguste Baum <auguste.apple@gmail.com>

mrobinson commented on components/script/dom/globalscope.rs:
```diff
@@ -731,6 +731,13 @@ impl FileListener {
     }
 }
 
+fn now() -> u64 {
```
I think it would be useful to indicate the units in the function name so something like `get_timestamp_in_ms`.

mrobinson commented on components/script/dom/htmlmediaelement.rs:
```diff
@@ -301,6 +301,10 @@ impl From<MediaStreamOrBlob> for SrcObject {
     }
 }
 
+fn now() -> SystemTime {
+    SystemTime::now()
+}
+
```
I think in this case you should omit the helper function entirely as it just calls a single function. It's more useful for the reader to know what this does.

mrobinson commented on components/script/script_runtime.rs:
```diff
@@ -761,22 +760,26 @@ unsafe extern "C" fn gc_slice_callback(
         }),
         GCProgress::GC_SLICE_END => GC_SLICE_START.with(|start| {
-            let dur = now() - start.get().unwrap();
+            let dur = SystemTime::now()
+                .duration_since(start.get().unwrap())
+                .unwrap();
```
I think we should try to not panic here, so please provide defaults if we cannot unwrap successfully.

augustebaum replied:
Changed the last `unwrap` to `unwrap_or_default`, for lack of a smarter solution.

mrobinson commented on components/script/script_runtime.rs:
```diff
@@ -761,22 +760,26 @@ unsafe extern "C" fn gc_slice_callback(
         }),
         GCProgress::GC_CYCLE_END => GC_CYCLE_START.with(|start| {
-            let dur = now() - start.get().unwrap();
+            let dur = SystemTime::now()
+                .duration_since(start.get().unwrap())
+                .unwrap();
```
The same comment here.

augustebaum replied:
See [previous conversation](https://github.com/servo/servo/pull/30639#discussion_r1375412348).

mrobinson commented on components/script/script_thread.rs:
```diff
@@ -214,9 +214,9 @@ struct InProgressLoad {
     #[no_trace]
     origin: MutableOrigin,
-    /// Timestamp reporting the time when the browser started this load.
+    /// Timestamp reporting the time in milliseconds when the browser started this load.
```
Great comment update!

mrobinson commented on components/script/script_thread.rs:
```diff
@@ -258,7 +261,7 @@ impl InProgressLoad {
             url: url,
             origin: origin,
-            navigation_start: (current_time.sec * 1000 + current_time.nsec as i64 / 1000000) as u64,
+            navigation_start: navigation_start,
```
```suggestion
            navigation_start,
```

augustebaum replied:
I contemplated doing this but it seems this would be better dealt with in a big `fmt` change.

sagudev replied:
We could enforce it with [use_field_init_shorthand](https://rust-lang.github.io/rustfmt/?version=v1.4.36&search=use_field_init_shorthand#use_field_init_shorthand).

mrobinson commented on components/script/script_thread.rs:
```diff
@@ -4059,11 +4068,8 @@ impl Drop for ScriptThread {
-        .strftime("%m/%d/%Y %H:%M:%S")
-        .unwrap()
-        .to_string()
+    let tm: Duration = tm.duration_since(SystemTime::UNIX_EPOCH).unwrap();
+    let dt = NaiveDateTime::from_timestamp_opt(tm.as_secs() as i64, tm.subsec_nanos()).unwrap();
+    let dt: DateTime<Local> = Local.from_utc_datetime(&dt);
+    dt.format("%m/%d/%Y %H:%M:%S").to_string()
```
Maybe we can improve the name of these variables while we are here. Maybe this?

- `tm` –> `current_time`
- `dt` –> `utc_time`
- `dt` –> `local_time`

Please also properly handle when the unwrap would fail here.

augustebaum replied:
Yeah I've been putting `unwrap`s all over the place because I have no visibility on the overall functionality. Would you have any idea how to deal with this specific one?

augustebaum replied:
See [previous comment](https://github.com/servo/servo/pull/30639#discussion_r1375412348).

mrobinson reviewed (changes requested):
Nicely done! I have a few comments below, but generally speaking this looks great.

augustebaum commented:
I'm nearly done --- only have `script/dom/htmlformelement.rs` left, but I'm getting a gnarly compilation error. See [this chat entry](https://servo.zulipchat.com/#narrow/stream/263398-general/topic/.2E.2E.2E.20Another.20compilation.20error.20involving.20.60dom_struct.60/near/399136903). Would appreciate any support!

augustebaum committed:
Replace `time` with `std::time` in `script/dom/htmlformelement.rs`

Signed-off-by: Auguste Baum <auguste.apple@gmail.com>

augustebaum committed:
Replace `time` with `std::time` in `script/dom/htmlformelement.rs`

Signed-off-by: Auguste Baum <auguste.apple@gmail.com>

mrobinson commented:
Looks like this is a flake in the float tests...frustrating and they should probably be switched away from a fuzzing approach to a traditional test framework. I'm resending this to the MQ.

augustebaum commented:
@mrobinson Thanks for the review! Please consider resolving the above conversations as I'm interested in your opinion on these.

mrobinson commented:
It seems that some animation and performance tests are failing, likely because the new code is returning different results that the old code. One way to debug this is to compare the two times (via the old APIs and the new ones) via print statements to see where they are different.

Failures:
- https://github.com/servo/servo/runs/18185432560
- https://github.com/servo/servo/runs/18185642035

augustebaum commented on components/script/animation_timeline.rs:
```diff
@@ -22,7 +23,10 @@ impl AnimationTimeline {
+            current_value: SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .unwrap_or_default()
+                .as_secs() as f64,
```
Quite right, sorry I missed this.

augustebaum commented on components/script/animation_timeline.rs:
```diff
@@ -39,7 +43,10 @@ impl AnimationTimeline {
+        self.current_value = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap_or_default()
+            .as_secs() as f64;
```
ditto

augustebaum commented:
Some of these test failures might be linked to the fact that [`std::time::SystemTime` is not necessarily monotonic](https://doc.rust-lang.org/std/time/struct.SystemTime.html). Will need to investigate this...

sagudev commented:
Where possible use https://doc.rust-lang.org/std/time/struct.Instant.html

mrobinson commented:
The new set of results looks a lot better! I think this is getting quite close. I would also rebase this as at least one of these failures should be fixed in the latest from the main branch.

augustebaum committed:
Replace `time` with `chrono` in `script/animation_timeline`

Signed-off-by: Auguste Baum <auguste.apple@gmail.com>

augustebaum commented:
I'm reworking the changes and there is something I find strange.

The `console_timers` attribute takes `u64` as values:
https://github.com/servo/servo/blob/f2270715861d577389fe6349a32ac483199d09f2/components/script/dom/globalscope.rs#L207

The first time they're used it's to put in a timestamp:
https://github.com/servo/servo/blob/f2270715861d577389fe6349a32ac483199d09f2/components/script/dom/globalscope.rs#L2272

whereas the second time it's to put in the elapsed time since the timestamp:
https://github.com/servo/servo/blob/f2270715861d577389fe6349a32ac483199d09f2/components/script/dom/globalscope.rs#L2284

Ideally `console_timers` would hold `Duration`s, but that doesn't fit with its first usage where `Instant` would be more fitting. I'll keep using `u64` but this inconsistency might bite someone down the line and should be dealt with.

augustebaum commented:
Kind of stuck here, in part because of https://github.com/servo/servo/issues/30714 but also just out of my depth. Is there any way to get support on this?

augustebaum commented:
Pinging just in case. Let me know if there is anything I can try.

mrobinson commented:
I think a good approach for debugging this is:

1. Confirm that at least one of the failing tests fails locally (maybe try to pick the simplest one).
2. Figure out what part of your change you need to revert to get it passing again.
3. Focus on that small part of your change to see what might be going wrong.

mrobinson committed:
Increase precision of animation timeline

mrobinson commented:
Looks like there is at least one more issue to address.

mrobinson committed:
Some fixes

Use Instant a bit more and stop using chrono. Do not transition
`navigation_start_precise` to Instant yet as we need to coordinate this
across all crates.

mrobinson commented:
I had to make some changes to get this working namely, fixing some precision issues in the animation timeline as well as not making the changes for `navigation_start_precise`. I think that in the former case, we will need to make these changes across the entire source code all at once.
